src/observers/qos_correction.hpp

```diff
@@ -26,73 +26,99 @@ namespace mesos {
+ * produces empty Corrections).
  */
 class QoSCorrectionObserver : public SyncConsumer<Contentions>,
                               public Consumer<ResourceUsage>,
-                              public Producer<QoSCorrections> {
+                              public Producer<QoSCorrections>,
+                              public Producer<Contentions> {
  public:
   explicit QoSCorrectionObserver(
       Consumer<QoSCorrections>* _consumer,
       uint64_t _contentionProducents,
       ExecutorAgeFilter* _ageFilter = new ExecutorAgeFilter(),
-      RevocationStrategy* _revStrategy = nullptr)
+      RevocationStrategy* _revStrategy =
+          new SeniorityStrategy(SerenityConfig()),
+      uint32_t _cooldownIterations = strategy::DEFAULT_CONTENTION_COOLDOWN,
+      const Tag& _tag = Tag(QOS_CONTROLLER, NAME))
     : SyncConsumer<Contentions>(_contentionProducents),
       Producer<QoSCorrections>(_consumer),
-      currentContentions(None()),
-      currentUsage(None()),
-      revStrategy(_revStrategy),
-      ageFilter(_ageFilter) {
-    if (revStrategy == nullptr) {
-      revStrategy = new SeniorityStrategy(SerenityConfig());
-    }
-  }
+      contentions(None()),
+      usage(None()),
+      revocationStrategy(_revStrategy),
+      executorAgeFilter(_ageFilter),
+      cooldownIterations(_cooldownIterations),
+      tag(_tag) {}
 
   ~QoSCorrectionObserver();
 
-  Try<Nothing> _syncConsume(const std::vector<Contentions> products) override;
-
+  Try<Nothing> syncConsume(const std::vector<Contentions> products) override;
   Try<Nothing> consume(const ResourceUsage& usage) override;
 
-  static bool compareSeverity(
-      const Contention& first, const Contention& second) {
-    if (!second.has_severity())
-      return true;
-    if (!first.has_severity())
-      return false;
-
-    return (first.severity() > second.severity());
+  template <typename T>
+  Try<Nothing> addConsumer(T* consumer) {
+    return Producer<T>::addConsumer(consumer);
   }
 
-  static bool compareCpuAllocated(
-      const ResourceUsage_Executor& first,
-      const ResourceUsage_Executor& second) {
-    // TODO(bplotka) decide if we want to validate fields here.
-    Resources secondAllocation(second.allocated());
-    if (secondAllocation.cpus().isNone())
-      return true;
-
-    Resources firstAllocation(first.allocated());
-    if (firstAllocation.cpus().isNone())
-      return false;
-
-    return (firstAllocation.cpus().get() > secondAllocation.cpus().get());
-  }
-
-  static constexpr const char* NAME = "CorrectionObserver: ";
-  static constexpr const char* name = "[SerenityQoS] CorrectionObserver: ";
+  static constexpr const char* NAME = "QoSCorrectionObserver: ";
 
  protected:
-  Option<Contentions> currentContentions;
-  Option<ResourceUsage> currentUsage;
-  RevocationStrategy* revStrategy;
-  ExecutorAgeFilter* ageFilter;
+  Try<Nothing> qosCorrectionStateMachine();
+  void noContentionsReceived();
```

Comment:
What about adding one newline between all function declarations in class? Look quite messy now.